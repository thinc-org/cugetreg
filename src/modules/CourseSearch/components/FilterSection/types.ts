export interface FilterSectionProps {
  open: boolean
  setOpen: React.Dispatch<React.SetStateAction<boolean>>
}
