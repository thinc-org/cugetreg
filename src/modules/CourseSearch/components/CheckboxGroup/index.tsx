import { Checkbox, CheckboxProps, FormControlLabel, Stack, Typography } from '@mui/material'
import React from 'react'

import { GeneralChipKey } from '@/common/components/Chips/config'
import { useSearchCourseQueryParams } from '@/modules/CourseSearch/hooks/useSearchCourseQueryParams'

export interface EnhancedCheckBoxProps extends CheckboxProps {
  label: string
  value: GeneralChipKey
}

export interface CheckboxGroupProps {
  title: string
  id: string
  checkboxes: EnhancedCheckBoxProps[]
  log: (_: unknown, value: string) => void
}

export const CheckboxGroup: React.FC<CheckboxGroupProps> = ({ title, checkboxes, log, id }) => {
  const { searchCourseQueryParams } = useSearchCourseQueryParams()

  const hasChecked = (tag: GeneralChipKey) => {
    const genEdTags: string[] = searchCourseQueryParams.filter?.genEdTypes ?? []
    const dayOfWeeks: string[] = searchCourseQueryParams.filter?.dayOfWeeks ?? []

    return genEdTags.includes(tag) || dayOfWeeks.includes(tag)
  }

  return (
    <Stack>
      <Typography variant="button" color="primaryRange.100">
        {title}
      </Typography>
      {checkboxes.map(({ label, ...checkbox }) => (
        <FormControlLabel
          key={label}
          onClick={() => log(null, label || '')}
          control={
            <Checkbox
              id={id}
              color="primary"
              sx={{ fontFamily: 'Prompt' }}
              checked={hasChecked(checkbox.value)}
              {...checkbox}
            />
          }
          label={<Typography variant="subtitle1">{label}</Typography>}
        />
      ))}
    </Stack>
  )
}
