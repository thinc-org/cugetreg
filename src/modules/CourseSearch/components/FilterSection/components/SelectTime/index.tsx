import { MenuItem, Select, Stack, Typography, makeStyles, Checkbox, FormControlLabel, Box } from '@material-ui/core'

import { useSelectTime } from './hooks/useSelectTime'

const useStyles = makeStyles((theme) => ({
  title: {
    color: theme.palette.primaryRange[100],
  },
}))

interface SelectTimeProps {
  log: (_: unknown, value: string) => void
}

export const SelectTime = ({ log }: SelectTimeProps) => {
  const {
    selectedStartTime,
    selectedEndTime,
    startTimeChoices,
    endTimeChoices,
    onStartTimeChange,
    onEndTimeChange,
    checked,
    onCheckboxChange,
  } = useSelectTime('06:00', '22:00')
  const classes = useStyles()

  return (
    <Stack>
      <Typography variant="button" className={classes.title}>
        เวลาเรียน
      </Typography>
      <FormControlLabel
        key="periodRange"
        onClick={() => log(null, 'periodRange')}
        sx={{
          display: 'flex',
          alignItems: 'flex-start',
        }}
        control={
          <Checkbox
            color="primary"
            name="เวลาเรียน"
            sx={{ fontFamily: 'Prompt' }}
            checked={checked}
            onChange={onCheckboxChange}
          />
        }
        label={
          <Box>
            <Box mb={2} display="flex" alignItems="center" justifyContent="space-between">
              <Typography mr={2} variant="subtitle1">
                ในช่วง
              </Typography>
              <Select
                disabled={!checked}
                value={selectedStartTime}
                onChange={(e) => {
                  const startTime = e.target.value as string
                  onStartTimeChange(startTime)
                }}
                name="startTime"
              >
                {startTimeChoices.map((value) => (
                  <MenuItem key={value} value={value}>
                    {value}
                  </MenuItem>
                ))}
              </Select>
            </Box>
            <Box display="flex" alignItems="center" justifyContent="space-between">
              <Typography mr={2} variant="subtitle1">
                ถึง
              </Typography>
              <Select
                disabled={!checked}
                value={selectedEndTime}
                onChange={(e) => {
                  const endTime = e.target.value as string
                  onEndTimeChange(endTime)
                }}
                name="endTime"
              >
                {endTimeChoices.map((value) => (
                  <MenuItem key={value} value={value}>
                    {value}
                  </MenuItem>
                ))}
              </Select>
            </Box>
          </Box>
        }
      />
    </Stack>
  )
}
