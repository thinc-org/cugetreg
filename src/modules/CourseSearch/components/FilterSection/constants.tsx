import { DayOfWeekEnum, GenEdTypeEnum } from '@thinc-org/chula-courses'

import { DayChipKey, GenEdChipKey, OtherChipKey } from '@/common/components/Chips/config'
import { CreateCheckbox } from '@/modules/CourseSearch/components/FilterSection/hooks'

export const createGenEdCheckboxes: CreateCheckbox<GenEdChipKey>[] = [
  {
    label: 'หมวดวิทย์',
    value: GenEdTypeEnum.SC,
  },
  {
    label: 'หมวดสังคม',
    value: GenEdTypeEnum.SO,
  },
  {
    label: 'หมวดมนุษย์',
    value: GenEdTypeEnum.HU,
  },
  {
    label: 'หมวดสหฯ',
    value: GenEdTypeEnum.IN,
  },
  {
    label: 'ไม่ใช่ GenEd',
    value: GenEdTypeEnum.NO,
  },
]

export const createDayOfWeekCheckboxes: CreateCheckbox<DayChipKey>[] = [
  {
    label: 'วันจันทร์',
    value: DayOfWeekEnum.Monday,
  },
  {
    label: 'วันอังคาร',
    value: DayOfWeekEnum.Tuesday,
  },
  {
    label: 'วันพุธ',
    value: DayOfWeekEnum.Wednesday,
  },
  {
    label: 'วันพฤหัสบดี',
    value: DayOfWeekEnum.Thursday,
  },
  {
    label: 'วันศุกร์',
    value: DayOfWeekEnum.Friday,
  },
  {
    label: 'วันเสาร์',
    value: DayOfWeekEnum.Saturday,
  },
  {
    label: 'วันอาทิตย์',
    value: DayOfWeekEnum.Sunday,
  },
]

export const createSpecialCheckboxes: CreateCheckbox<OtherChipKey>[] = [
  {
    label: 'เวลาไม่ชนกับวิชาที่เลือก',
    value: 'noConflict',
  },
]
