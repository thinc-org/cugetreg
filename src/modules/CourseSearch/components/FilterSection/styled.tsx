import { css } from '@emotion/react'
import styled from '@emotion/styled'
import { Button as MuiButton, DialogContent as MuiDialogContent, Paper } from '@mui/material'

export const DialogContent = styled(MuiDialogContent)`
  padding: ${({ theme }) => theme.spacing(4)};
`

export const Box = styled.div`
  display: flex;
  justify-content: space-between;
  div:first-of-type {
    ${({ theme }) => {
      return css`
        margin-right: ${theme.spacing(12)};
        ${theme.breakpoints.down('sm')} {
          margin-right: ${theme.spacing(2)};
        }
      `
    }};
  }
`

export const StickyPaper = styled(Paper)<{ hasTags: boolean }>`
  display: flex;
  position: sticky;
  width: fit-content;
  height: fit-content;
  top: ${({ hasTags }) => (hasTags ? '125px' : '101px')};
  max-height: calc(100vh - ${({ hasTags }) => (hasTags ? '125px' : '101px')} - 24px);
`

export const Button = styled(MuiButton)`
  margin-top: ${({ theme }) => theme.spacing(2)};
`
