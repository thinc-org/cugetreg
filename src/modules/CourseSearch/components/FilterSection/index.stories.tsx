import { Meta, Story } from '@storybook/react/types-6-0'
import React from 'react'

import { FilterSection } from '@/modules/CourseSearch/components/FilterSection'
import { FilterSectionProps } from '@/modules/CourseSearch/components/FilterSection/types'

type StoryProps = FilterSectionProps

export default {
  title: 'Component/FilterBar',
  component: FilterSection,
  argTypes: {},
} as Meta<StoryProps>

export const Default: Story<StoryProps> = (args) => {
  return <FilterSection {...args} />
}

Default.args = {}
