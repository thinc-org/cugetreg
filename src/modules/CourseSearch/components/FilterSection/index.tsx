import { useTheme } from '@emotion/react'
import { DialogContent, Grid } from '@material-ui/core'
import useMediaQuery from '@material-ui/core/useMediaQuery'
import useGoogleOptimize from '@react-hook/google-optimize'

import { DayChipKey, GenEdChipKey } from '@/common/components/Chips/config'
import { ResponsiveDialog } from '@/common/components/ResponsiveDialog'
import { Analytics } from '@/common/context/Analytics/components/Analytics'
import { DAY_FILTER, GENED_FILTER, PERIOD_RANGE_FILTER } from '@/common/context/Analytics/constants'
import { CheckboxGroup } from '@/modules/CourseSearch/components/CheckboxGroup'
import {
  createGenEdCheckboxes,
  createDayOfWeekCheckboxes, // createSpecialCheckboxes,
} from '@/modules/CourseSearch/components/FilterSection/constants'
import { FilterSectionProps } from '@/modules/CourseSearch/components/FilterSection/types'
import { google_optimize_filter_order } from '@/utils/env'

import { useHasTags } from '../TagList'
import { SelectTime } from './components/SelectTime'
import { useFilterBar } from './hooks/useFilterBar'
import { Button, StickyPaper, Box } from './styled'

export const FilterSection: React.FC<FilterSectionProps> = ({ open, handleClose }) => {
  const { checkboxes: genEdCheckboxes } = useFilterBar<GenEdChipKey>(createGenEdCheckboxes, 'genEdTypes')
  const { checkboxes: dayOfWeekCheckboxes } = useFilterBar<DayChipKey>(createDayOfWeekCheckboxes, 'dayOfWeeks')
  // const { checkboxes: specialCheckboxes } = useFilterBar(createSpecialCheckboxes)

  const hasTags = useHasTags()

  const theme = useTheme()
  const match = useMediaQuery(theme.breakpoints.up('sm'))
  const isExperimentOrder = useGoogleOptimize(google_optimize_filter_order, [false, true])

  return match ? (
    open ? (
      <StickyPaper hasTags={hasTags} variant="outlined">
        <Grid container spacing={4}>
          <Grid item sx={{ order: isExperimentOrder ? 3 : 1 }}>
            <Analytics elementName={GENED_FILTER}>
              {({ log }) => <CheckboxGroup log={log} title="หมวดหมู่ GenEd" checkboxes={genEdCheckboxes} />}
            </Analytics>
          </Grid>
          <Grid item sx={{ order: isExperimentOrder ? 1 : 2 }}>
            <Analytics elementName={DAY_FILTER}>
              {({ log }) => <CheckboxGroup log={log} title="วันในสัปดาห์" checkboxes={dayOfWeekCheckboxes} />}
            </Analytics>
          </Grid>
          {/* <CheckboxGroup title="แสดงผลพิเศษ" checkboxes={specialCheckboxes} /> */}
          <Grid item sx={{ order: isExperimentOrder ? 2 : 3 }}>
            <Analytics elementName={PERIOD_RANGE_FILTER}>{({ log }) => <SelectTime log={log} />}</Analytics>
          </Grid>
        </Grid>
      </StickyPaper>
    ) : null
  ) : (
    <ResponsiveDialog open={open} onClose={handleClose}>
      <DialogContent>
        <Box>
          <Analytics elementName={GENED_FILTER}>
            {({ log }) => <CheckboxGroup log={log} title="หมวดหมู่ GenEd" checkboxes={genEdCheckboxes} />}
          </Analytics>
          <Analytics elementName={DAY_FILTER}>
            {({ log }) => <CheckboxGroup log={log} title="วันในสัปดาห์" checkboxes={dayOfWeekCheckboxes} />}
          </Analytics>
        </Box>
        {/* <CheckboxGroup title="แสดงผลพิเศษ" checkboxes={specialCheckboxes} /> */}
        <Button color="primary" variant="outlined" fullWidth onClick={handleClose}>
          เลือกตัวกรอง
        </Button>
      </DialogContent>
    </ResponsiveDialog>
  )
}
