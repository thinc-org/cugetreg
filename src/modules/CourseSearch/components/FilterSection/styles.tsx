import { css } from '@emotion/react'
import styled from '@emotion/styled'
import { Button as MuiButton, DialogContent as MuiDialogContent, Paper } from '@material-ui/core'

export const DialogContent = styled(MuiDialogContent)`
  padding: ${({ theme }) => theme.spacing(4)};
`

export const Box = styled.div`
  display: flex;
  justify-content: space-between;
  div:first-of-type {
    ${({ theme }) => {
      return css`
        margin-right: ${theme.spacing(12)};
        ${theme.breakpoints.down('sm')} {
          margin-right: ${theme.spacing(2)};
        }
      `
    }};
  }
`

export const StickyPaper = styled(Paper)<{ hasTags: boolean }>`
  min-width: 280px;
  position: sticky;
  top: ${({ hasTags }) => (hasTags ? '125px' : '101px')};

  padding: ${({ theme }) => theme.spacing(4)};
  overflow-y: auto;
  height: min-content;
  max-height: calc(100vh - ${({ hasTags }) => (hasTags ? '125px' : '101px')});
`

export const Button = styled(MuiButton)`
  margin-top: ${({ theme }) => theme.spacing(2)};
`
