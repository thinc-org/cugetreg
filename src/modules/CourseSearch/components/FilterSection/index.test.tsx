import { DayOfWeekEnum, GenEdTypeEnum } from '@thinc-org/chula-courses'
import { shallow } from 'enzyme'

import { ResponsiveDialog } from '@/common/components/ResponsiveDialog'
import { Analytics } from '@/common/context/Analytics/components/Analytics'
import { DAY_FILTER, GENED_FILTER } from '@/common/context/Analytics/constants'
import { Button, StickyPaper } from '@/modules/CourseSearch/components/FilterSection/styles'

describe('FilterSection', () => {
  const mockUseMediaQuery = jest.fn()
  const mockUseFilterBar = jest.fn(() => ({
    checkboxes: [],
  }))
  const MOCK_HASTAG = true
  const mockUseHasTags = jest.fn(() => MOCK_HASTAG)

  const MOCK_QUERY = 'query'
  const mockUseTheme = jest.fn(() => ({
    breakpoints: {
      up: jest.fn(() => MOCK_QUERY),
    },
  }))
  jest.doMock('@material-ui/core/useMediaQuery', () => mockUseMediaQuery)
  jest.doMock('./hooks/useFilterBar', () => ({ useFilterBar: mockUseFilterBar }))
  jest.doMock('../TagList', () => ({ useHasTags: mockUseHasTags }))
  jest.doMock('@emotion/react', () => ({ ...jest.requireActual('@emotion/react'), useTheme: mockUseTheme }))

  afterEach(() => {
    jest.clearAllMocks()
  })

  it.each`
    matchSmallScreen | isExpandFilter
    ${true}          | ${true}
    ${true}          | ${false}
    ${false}         | ${true}
    ${false}         | ${false}
  `(
    'Should matched snapshot correctly when matchSmallScreen=$matchSmallScreen and isExpandFilter=$isExpandFilter',
    async ({ matchSmallScreen, isExpandFilter }) => {
      const { FilterSection } = await import('.')
      const mockHandleClose = jest.fn()
      mockUseMediaQuery.mockReturnValue(matchSmallScreen)

      const wrapper = shallow(<FilterSection open={isExpandFilter} handleClose={mockHandleClose} />)

      expect(mockUseMediaQuery).toBeCalledWith(MOCK_QUERY)
      expect(mockUseTheme).toBeCalledTimes(1)
      expect(mockUseHasTags).toBeCalledTimes(1)
      expect(mockUseFilterBar).toBeCalledTimes(2)
      expect(mockUseFilterBar).toHaveBeenNthCalledWith(
        1,
        [
          {
            label: 'หมวดวิทย์',
            value: GenEdTypeEnum.SC,
          },
          {
            label: 'หมวดสังคม',
            value: GenEdTypeEnum.SO,
          },
          {
            label: 'หมวดมนุษย์',
            value: GenEdTypeEnum.HU,
          },
          {
            label: 'หมวดสหฯ',
            value: GenEdTypeEnum.IN,
          },
          {
            label: 'ไม่ใช่ GenEd',
            value: GenEdTypeEnum.NO,
          },
        ],
        'genEdTypes'
      )
      expect(mockUseFilterBar).toHaveBeenNthCalledWith(
        2,
        [
          {
            label: 'วันจันทร์',
            value: DayOfWeekEnum.Monday,
          },
          {
            label: 'วันอังคาร',
            value: DayOfWeekEnum.Tuesday,
          },
          {
            label: 'วันพุธ',
            value: DayOfWeekEnum.Wednesday,
          },
          {
            label: 'วันพฤหัสบดี',
            value: DayOfWeekEnum.Thursday,
          },
          {
            label: 'วันศุกร์',
            value: DayOfWeekEnum.Friday,
          },
          {
            label: 'วันเสาร์',
            value: DayOfWeekEnum.Saturday,
          },
          {
            label: 'วันอาทิตย์',
            value: DayOfWeekEnum.Sunday,
          },
        ],
        'dayOfWeeks'
      )

      if (!matchSmallScreen) {
        expect(wrapper.find(ResponsiveDialog).prop('onClose')).toBe(mockHandleClose)
        expect(wrapper.find(Button).prop('onClick')).toBe(mockHandleClose)
      }
      if (matchSmallScreen && isExpandFilter) {
        expect(wrapper.find(StickyPaper).prop('hasTags')).toBe(MOCK_HASTAG)
      }
      expect(wrapper).toMatchSnapshot()
    }
  )

  it.each`
    matchSmallScreen | isExpandFilter
    ${true}          | ${true}
    ${false}         | ${true}
    ${false}         | ${false}
  `(
    'Should render CheckboxGroup of GenEds correctly when when matchSmallScreen=$matchSmallScreen and isExpandFilter=$isExpandFilter',
    async ({ matchSmallScreen, isExpandFilter }) => {
      const { FilterSection } = await import('.')
      const mockHandleClose = jest.fn()
      mockUseMediaQuery.mockReturnValue(matchSmallScreen)

      const wrapper = shallow(<FilterSection open={isExpandFilter} handleClose={mockHandleClose} />)
        .find(Analytics)
        .findWhere((n) => n.prop('elementName') === GENED_FILTER)
        .renderProp('children')({ log: () => {} })
      expect(wrapper).toMatchSnapshot()
    }
  )

  it.each`
    matchSmallScreen | isExpandFilter
    ${true}          | ${true}
    ${false}         | ${true}
    ${false}         | ${false}
  `(
    'Should render CheckboxGroup of days correctly when matchSmallScreen=$matchSmallScreen and isExpandFilter=$isExpandFilter',
    async ({ matchSmallScreen, isExpandFilter }) => {
      const { FilterSection } = await import('.')
      const mockHandleClose = jest.fn()
      mockUseMediaQuery.mockReturnValue(matchSmallScreen)

      const wrapper = shallow(<FilterSection open={isExpandFilter} handleClose={mockHandleClose} />)
        .find(Analytics)
        .findWhere((n) => n.prop('elementName') === DAY_FILTER)
        .renderProp('children')({ log: () => {} })
      expect(wrapper).toMatchSnapshot()
    }
  )
})

shallow
