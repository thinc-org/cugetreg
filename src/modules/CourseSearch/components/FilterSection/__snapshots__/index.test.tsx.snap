// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSection Should matched snapshot correctly when matchSmallScreen=false and isExpandFilter=false 1`] = `
<ResponsiveDialog
  onClose={[MockFunction]}
  open={false}
>
  <ForwardRef(DialogContent)>
    <Styled(div)>
      <Analytics
        elementName="gened_search_filter"
      >
        <Component />
      </Analytics>
      <Analytics
        elementName="day_search_filter"
      >
        <Component />
      </Analytics>
    </Styled(div)>
    <Styled(Component)
      color="primary"
      fullWidth={true}
      onClick={[MockFunction]}
      variant="outlined"
    >
      เลือกตัวกรอง
    </Styled(Component)>
  </ForwardRef(DialogContent)>
</ResponsiveDialog>
`;

exports[`FilterSection Should matched snapshot correctly when matchSmallScreen=false and isExpandFilter=true 1`] = `
<ResponsiveDialog
  onClose={[MockFunction]}
  open={true}
>
  <ForwardRef(DialogContent)>
    <Styled(div)>
      <Analytics
        elementName="gened_search_filter"
      >
        <Component />
      </Analytics>
      <Analytics
        elementName="day_search_filter"
      >
        <Component />
      </Analytics>
    </Styled(div)>
    <Styled(Component)
      color="primary"
      fullWidth={true}
      onClick={[MockFunction]}
      variant="outlined"
    >
      เลือกตัวกรอง
    </Styled(Component)>
  </ForwardRef(DialogContent)>
</ResponsiveDialog>
`;

exports[`FilterSection Should matched snapshot correctly when matchSmallScreen=true and isExpandFilter=false 1`] = `""`;

exports[`FilterSection Should matched snapshot correctly when matchSmallScreen=true and isExpandFilter=true 1`] = `
<Styled(Component)
  hasTags={true}
  variant="outlined"
>
  <ForwardRef(Stack)
    spacing={4}
  >
    <Analytics
      elementName="gened_search_filter"
    >
      <Component />
    </Analytics>
    <Analytics
      elementName="day_search_filter"
    >
      <Component />
    </Analytics>
  </ForwardRef(Stack)>
</Styled(Component)>
`;
